lib/Sema/TypeCheckAvailability.cpp

```diff
@@ -491,12 +504,72 @@ class TypeRefinementContextBuilder : private ASTWalker {
 
     while (!DeclBodyContextStack.empty() &&
            DeclBodyContextStack.back().Decl == D) {
+      // All pending body TRCs should have been consumed.
+      assert(DeclBodyContextStack.back().BodyTRCs.empty());
       DeclBodyContextStack.pop_back();
     }
 
     return Action::Continue();
   }
 
+  bool shouldBuildLazyContextForDecl(Decl *D) {
+    // Skip functions that have unparsed bodies on an initial descent to avoid
+    // eagerly parsing bodies unnecessarily.
+    if (auto *afd = dyn_cast<AbstractFunctionDecl>(D)) {
+      if (afd->hasBody() && !afd->isBodySkipped() &&
+          !afd->getBody(/*canSynthesize=*/false))
+        return true;
+    }
+
+    // Pattern binding declarations may have attached property wrappers that
+    // get expanded from macros attached to the parent declaration. We must
+    // not eagerly expand the attached property wrappers to avoid request
+    // cycles.
+    if (auto *pattern = dyn_cast<PatternBindingDecl>(D)) {
+      if (auto firstVar = pattern->getAnchoringVarDecl(0)) {
+        // FIXME: We could narrow this further by detecting whether there are
+        // any macro expansions required to visit the CustomAttrs of the var.
+        if (firstVar->hasInitialValue() && !firstVar->isInitExposedToClients())
+          return true;
+      }
+    }
+
+    return false;
```

Comment:
Could we choose to always be lazy and still have things work?

Reply:
I think so, but the expense would be allocating a TRC node for every declaration, regardless of whether it introduces anything interesting availability-wise. I suspect that would have noticeable costs in both memory use and in query performance. But perhaps it's worth experimenting with to see just how much.